Code/GraphMol/Abbreviations/Abbreviations.cpp

```diff
@@ -0,0 +1,256 @@
+      toRemove.set(pr.second);
+      // if there's a molecule associated with the match, check to see if
+      // additional bonds need to be formed
+      if (amatch.abbrev.mol &&
+          mol.getAtomWithIdx(pr.second)->getDegree() >
+              amatch.abbrev.mol->getAtomWithIdx(pr.first)->getDegree()) {
+        for (const auto& nbri : boost::make_iterator_range(
+                 mol.getAtomNeighbors(mol.getAtomWithIdx(pr.second)))) {
+          const auto& nbr = mol[nbri];
+          auto nbrIdx = nbr->getIdx();
+          // if this neighbor isn't in the match:
+          if (!std::any_of(amatch.match.begin(), amatch.match.end(),
+                           [&](const std::pair<int, int>& tpr) {
+                             return tpr.second == rdcast<int>(nbrIdx);
+                           })) {
+            mol.addBond(nbrIdx, connectIdx, Bond::BondType::SINGLE);
+          }
+        }
+      }
+    }
+    // make connections between any extraAttachAtoms and the connection point
+    for (auto oaidx : amatch.abbrev.extraAttachAtoms) {
+      mol.addBond(oaidx, connectIdx, Bond::BondType::SINGLE);
+    }
+  }
+  for (unsigned int i = toRemove.size(); i > 0; --i) {
+    if (toRemove[i - 1]) {
+      mol.removeAtom(i - 1);
+    }
+  }
+}
+
+void labelMatches(RWMol& mol, const std::vector<AbbreviationMatch>& matches) {
+  for (const auto& amatch : matches) {
+    // throughout this remember that atom 0 in the match is the dummy
+    SubstanceGroup sg(&mol, "SUP");
+    sg.setProp("LABEL", amatch.abbrev.label);
+
+    for (unsigned int i = 1; i < amatch.match.size(); ++i) {
+      const auto& pr = amatch.match[i];
+      sg.addAtomWithIdx(pr.second);
+    }
+    auto bnd =
+        mol.getBondBetweenAtoms(amatch.match[0].second, amatch.match[1].second);
+    CHECK_INVARIANT(bnd, "bond to attachment point not found");
+    sg.addBondWithIdx(bnd->getIdx());
+    sg.addAttachPoint(amatch.match[1].second, amatch.match[0].second, "1");
+    addSubstanceGroup(mol, sg);
+  }
+}
+
+std::vector<AbbreviationMatch> findApplicableAbbreviationMatches(
+    const ROMol& mol, const std::vector<AbbreviationDefinition>& abbrevs,
+    double maxCoverage) {
+  std::vector<AbbreviationMatch> res;
+  auto nAtoms = mol.getNumAtoms();
+  if (!nAtoms || abbrevs.empty()) {
+    return res;
+  }
+
+  MolOps::fastFindRings(mol);
+
+  std::vector<AbbreviationMatch> tres;
+  boost::dynamic_bitset<> dummies(mol.getNumAtoms());
+  boost::dynamic_bitset<> firstAts(mol.getNumAtoms());
+  boost::dynamic_bitset<> covered(mol.getNumAtoms());
+
+  for (const auto& abbrev : abbrevs) {
+    CHECK_INVARIANT(abbrev.mol, "molecule is null");
+    if (maxCoverage > 0) {
+      unsigned int nDummies;
+      abbrev.mol->getProp(common_properties::numDummies, nDummies);
+      if (double(abbrev.mol->getNumAtoms() - nDummies) / nAtoms >=
+          maxCoverage) {
+        continue;
+      }
+    }
+    auto matches = SubstructMatch(mol, *abbrev.mol);
```

Comment:
What happens with overlapping matches or linkers?

Reply:
only the first one is used. That's the loop at line 140